### PR TITLE
Deprecate `LightningDataModule.on_save/load_checkpoint`

### LINKED ISSUE #11824: [RFC] Deprecate `CheckpointHooks` off `LightningDataModule`, `PrecisionPlugin`
## Proposed refactor

Following the addition and alignment of `Stateful` API in issue #11429, deprecate `CheckpointHooks` `on_save/load_checkpoint(checkpoint_dict)` off `LightningDataModule` and `PrecisionPlugin`.

### Motivation

Now that `LightningDataModule` and `PrecisionPlugin` can contribute and load their own component state with `Stateful` APIs, I don't think there's much use cases where these components need access to edit or peek at other component states in the checkpoint dictionary - could we deprecate `CheckpointHooks` off these components?

### Pitch

Deprecate `CheckpointHooks` off `LightningDataModule` and `PrecisionPlugin`.
Issue deprecation warning if `on_save/load_checkpoint` are overridden, remove in v1.8.

### Additional context

issue #11429 

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @rohitgr7 @ananthsub @ninginthecloud @tchaton

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #11824
Deprecate `LightningDataModule.on_save/load_checkpoint` hooks

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-8.py [hunk 1]
```diff
@@ -500,3 +500,35 @@ def on_before_accelerator_backend_setup(self, *args, **kwargs):
         " and will be removed in v1.8"
     ):
         trainer.fit(model)
+
+
+def test_v1_8_0_datamodule_checkpointhooks(tmpdir):
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply tests/deprecated_api/test_remove_1-8.py:505]: I was thinking this allows more proactive warnings - we don't have to wait for if and when checkpoint save/load happens, but rather warn whenever a datamodule with these hooks is being used

File: tests/core/test_datamodules.py [hunk 2]
```diff
@@ -223,7 +223,11 @@ def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
     )
 
     # fit model
-    trainer.fit(model, datamodule=dm)
+    with pytest.deprecated_call(
+        match="`LightningDataModule.on_save_checkpoint` was deprecated in"
```
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply tests/core/test_datamodules.py:227]: this is just to catch the warning here, the actual deprecation test for both `on_save/load_checkpoint` is in `test_remove_1-8.py`

[the remark this answers is not in the document]
Comment (carmocca) [reply tests/core/test_datamodules.py:227]: But below, the test calls `_restore_modules_and_callbacks(checkpoint_path)`. Shouldn't that be triggering a deprecation warning too?

File: tests/deprecated_api/test_remove_1-8.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/deprecated_api/test_remove_1-8.py:537]: I'm proposing this to reduce the test runtime and scope.

Relying on a `fit` call to test is a bit lazy and a pattern we've abused in the past and are trying to change now.

File: docs/source/extensions/datamodules.rst [hunk 3]
```diff
@@ -326,15 +326,35 @@ on_after_batch_transfer
...
+load_state_dict
+~~~~~~~~~~~~~~~~~~
+
+.. automethod:: pytorch_lightning.core.datamodule.LightningDataModule.load_state_dict
+    :noindex:
+
 on_save_checkpoint
 ~~~~~~~~~~~~~~~~~~
 
+.. warning:: ``on_save_checkpoint`` is deprecated in v1.6 and will be removed in v1.8.
```
Comment (carmocca) [inline docs/source/extensions/datamodules.rst:345]: Instead, I would not mention the deprecated methods in the docs.

Also, have you grepped the docs directory for any occurrences?

File: tests/deprecated_api/test_remove_1-8.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply tests/deprecated_api/test_remove_1-8.py:537]: oh I mean I purposely put the deprecation warnings inside `_run` so that the user can get more proactive warnings - gives up the specificity of conditioning and timing the warnings on if and when save/load actually happens, but this way warns whenever a datamodule with these hooks is being used.

lmk if less proactively and more specifically warning users only if and when save/load occurs is preferred over this though

File: tests/core/test_datamodules.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply tests/core/test_datamodules.py:227]: (currently no, bc of above conversation - warning in `_run` instead of exact save/load)

File: docs/source/extensions/datamodules.rst (hunk 3 above)
Comment (jjenniferdai) [reply docs/source/extensions/datamodules.rst]: > Also, have you grepped the docs directory for any occurrences?

Yep, this is it for datamodules. Will have a separate overall docs PR for `checkpointing.rst`

File: tests/core/test_datamodules.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply tests/core/test_datamodules.py:226]: (marking as resolved in favor of similar conversation above)

File: tests/deprecated_api/test_remove_1-8.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/deprecated_api/test_remove_1-8.py:693]: My request is that we write the test like this:

```python
def test_warning():
    trainer = Mock()
    
    trainer.datamodule = CustomBoringDataModuleSave()
    with pytest.deprecated_call(match=...)
        _check_datamodule_checkpoint_hooks(trainer)

    same for CustomBoringDataModuleLoad
    ...
```

Which would run much faster than 2 `trainer.fit` calls

**Review by carmocca — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-8.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjenniferdai) [reply tests/deprecated_api/test_remove_1-8.py:694]: ah got it

File: tests/deprecated_api/test_remove_1-8.py [hunk 4]
```diff
@@ -688,3 +689,29 @@ def on_load_checkpoint(self, checkpoint):
         " v1.6 and will be removed in v1.8. Use `load_state_dict` instead."
     ):
         trainer.fit(model)
+
+
+def test_v1_8_0_datamodule_checkpointhooks(tmpdir):
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:694]:
```suggestion
def test_v1_8_0_datamodule_checkpointhooks():
```

**Comment by jjenniferdai** [comment]
see issue discussion on whether or not to deprecate
PR discussion can just be about the PR

**Review by tchaton — APPROVED** [review]
LGTM !
